kong/router/utils.lua

```diff
@@ -0,0 +1,247 @@
+local constants = require("kong.constants")
+local hostname_type = require("kong.tools.utils").hostname_type
+local normalize = require("kong.tools.uri").normalize
+
+local type   = type
+local error  = error
+local find   = string.find
+local sub    = string.sub
+local byte   = string.byte
+
+
+local SLASH  = byte("/")
+
+
+local DEFAULT_HOSTNAME_TYPE = hostname_type("")
+
+
+local protocol_subsystem = constants.PROTOCOLS_WITH_SUBSYSTEM
+
+
+--[[
+Hypothesis
+----------
+
+Item size:        1024 bytes
+Max memory limit: 5 MiBs
+
+LRU size must be: (5 * 2^20) / 1024 = 5120
+Floored: 5000 items should be a good default
+--]]
+local MATCH_LRUCACHE_SIZE = 5e3
+
+
+local function sanitize_uri_postfix(uri_postfix)
+  if not uri_postfix or uri_postfix == "" then
+    return uri_postfix
+  end
+
+  if uri_postfix == "." or uri_postfix == ".." then
+    return ""
+  end
+
+  if sub(uri_postfix, 1, 2) == "./" then
+    return sub(uri_postfix, 3)
+  end
+
+  if sub(uri_postfix, 1, 3) == "../" then
+    return sub(uri_postfix, 4)
+  end
+
+  return uri_postfix
+end
+
+
+local function strip_uri_args(req_uri)
+  local idx = find(req_uri, "?", 2, true)
+  if idx then
+    req_uri = sub(req_uri, 1, idx - 1)
+  end
+
+  return normalize(req_uri, true)
+end
+
+
+local function check_select_params(req_method, req_uri, req_host, req_scheme,
+                                   src_ip, src_port,
+                                   dst_ip, dst_port,
+                                   sni, req_headers)
+  if req_method and type(req_method) ~= "string" then
+    error("method must be a string", 2)
+  end
+  if req_uri and type(req_uri) ~= "string" then
+    error("uri must be a string", 2)
+  end
+  if req_host and type(req_host) ~= "string" then
+    error("host must be a string", 2)
+  end
+  if req_scheme and type(req_scheme) ~= "string" then
+    error("scheme must be a string", 2)
+  end
+  if src_ip and type(src_ip) ~= "string" then
+    error("src_ip must be a string", 2)
+  end
+  if src_port and type(src_port) ~= "number" then
+    error("src_port must be a number", 2)
+  end
+  if dst_ip and type(dst_ip) ~= "string" then
+    error("dst_ip must be a string", 2)
+  end
+  if dst_port and type(dst_port) ~= "number" then
+    error("dst_port must be a number", 2)
+  end
+  if sni and type(sni) ~= "string" then
+    error("sni must be a string", 2)
+  end
+  if req_headers and type(req_headers) ~= "table" then
+    error("headers must be a table", 2)
+  end
+end
+
+
+local function debug_http_headers(var, header, match_t)
```

Comment:
```suggestion
local function add_debug_headers(var, header, match_t)
```

is probably better, `debug_http_headers` sounds a bit strange,

Reply:
done